docs/visual-studio-extensibility/nuget-api-in-visual-studio.md

```diff
@@ -23,35 +41,72 @@ As of NuGet 3.3+, NuGet exports the following
+
+- [`IVsSolutionRestoreService`](#ivssolutionrestoreservice-interface) (4.0+)
+- [`IVsSolutionRestoreService2`](#ivssolutionrestoreservice2-interface) (4.3+)
+- [`IVsSolutionRestoreService3`](#ivssolutionrestoreservice3-interface) (5.1+)
+- [`IVsSolutionRestoreService4`](#ivssolutionrestoreservice4-interface) (6.0+)
+- [`IVsSolutionRestoreStatusProvider`](#ivssolutionrestorestatusprovider-interface) (6.0+)
 
-    When installed, the package automatically sets the **Embed Interop Types** property of the assembly reference to **True**. This makes your code  resilient against version changes when users update to newer versions of NuGet.
+## Using NuGet Services
 
 > [!Warning]
-> Do not use any other types besides the public interfaces in your code, and do not reference any other NuGet assemblies, including `NuGet.Core.dll`.
+> Do not use any other types besides the public interfaces in your code, and do not reference any other NuGet assemblies, such as `NuGet.Protocol.dll`, `NuGet.Frameworks.dll`, and so on.
+
+In order to maximize the backwards compatibility promises we make, but also providing ourselves the flexibility to implement new features, performance improvements, and bug fixes in Visual Studio, we do not support the NuGet Client SDK being used in Visual Studio, and we do not provide binding redirects in `devenv.exe.config` to assemblies other than our VS extensibility contracts.
+
+If you would like a new NuGet related API in Visual Studio, please search [NuGet's Home repo](https://github.com/NuGet/Home/) and up vote any existing issues if you find a similar one. If you can't find an existing feature request to up vote, please create one.
+
+### Brokered Services
+
+1. Install the [`NuGet.VisualStudio.Contracts`](https://www.nuget.org/packages/NuGet.VisualStudio.Contracts/) package into your project, as well as [`Microsoft.VisualStudio.SDK`](https://www.nuget.org/packages/Microsoft.VisualStudio.SDK).
+
+1. Use the `IAsyncServiceProvider` to get Visual Studio's service broker, and use that to get NuGet's service.
+
+   ```cs
+   // Your AsyncPackage implements IAsyncServiceProvider
+   IAsyncServiceProvider asyncServiceProvider = this;
+   var brokeredServiceContainer = await asyncServiceProvider.GetServiceAsync<SVsBrokeredServiceContainer, IBrokeredServiceContainer>();
+   var serviceBroker = brokeredServiceContainer.GetFullAccessServiceBroker();
+   var nugetProjectService = await serviceBroker.GetProxyAsync<INuGetProjectService>(NuGetServices.NuGetProjectServiceV1);
+   ```
+
+1. When your code no longer needs NuGet's brokered service, dispose it. For example, if you only need NuGet's brokered service during a single method call, you can wrap it in a C#`using` statement:
+
+   ```cs
+   InstalledPackagesResult installedPackagesResult;
+   using (nugetProjectService as IDisposable)
```

Comment:
Wouldn't this throw if the service did not implement IDisposable?

Reply:
I haven't tested it myself. This was just the pattern suggested internally when I was learning about brokered services last year. In this specific case, we know that the service broker will definitely return a disposable object, even if the interface we cast to doesn't extend `IDisposable`.

I'm not sure why this pattern was used, I think it would be much better for package consumers if the interface extended `IDisposable`, but one side effect is that it will force the service implementation to implement `IDisposable`, even if it doesn't use any disposable objects itself. But that's not a big problem, we can have an empty `Dispose()` method. But changing our interface & service implementation is out of scope for this docs PR. This PR documents how it is now.

Reply:
Also to note, the brokered service does not return the object that our service factory creates. The service broker returns a dynamically created (with Castle?) proxy, which is disposable, which in turn uses StreamJsonRpc, or other technologies, to do RPC between the proxy and our service (hence it can work over Live Share, or locally).

But all this is implementation detail that we the NuGet team don't own, we just use, plus I believe it's too much detail and would confuse customers trying to get started with using our APIs.

Reply:
I am aware of what the service returns, but I didn';t think that you're always guaranteed to get a disposable proxy. 

Anyway, resolving.